File: 2021/day1/day1.cpp
#include <iostream>
#include <format>
#include "../util.h"

std::vector<int> _parse(const std::vector<std::string>& lines) {
  std::vector<int> data;
  for (auto& line : lines) {
    data.push_back(stoi(line));
  }
  return data;
}

int part1(const std::vector<int>& data) {
  int ret = 0;
  for (int i = 0; i < data.size() - 1; i++) {
    if (data[i] < data[i + 1]) {
      ret++;
    }
  }
  return ret;
}

int part2(const std::vector<int>& data) {
  int ret = 0;
  for (int i = 0; i < data.size() - 3; i++) {
    if (data[i] < data[i + 3]) {
      ret++;
    }
  }
  return ret;
}

int main(int argc, char** argv) {
  auto lines = read_lines("day1/input.txt");
  auto data = _parse(lines);

  {
    auto [result, ms] = measure_ms(part1, data);
    std::cout << std::format("part1 : {}\n - elapsed : {} ms\n", result, ms);
  }

  {
    auto [result, ms] = measure_ms(part2, data);
    std::cout << std::format("part2 : {}\n - elapsed : {} ms\n", result, ms);
  }

  return 0;
}

File: 2024/day1/day1.dart
import 'dart:io';

void part1(List<String> lines) {
  var left = lines.map((line) => int.parse(line.split('   ')[0])).toList()
    ..sort();
  var right = lines.map((line) => int.parse(line.split('   ')[1])).toList()
    ..sort();

  var ret = left
      .asMap()
      .entries
      .map((e) => (e.value - right[e.key]).abs())
      .reduce((a, b) => a + b);

  print(ret);
}

void part2(List<String> lines) {
  var left = lines.map((line) => int.parse(line.split('   ')[0])).toList();
  var right = lines.map((line) => int.parse(line.split('   ')[1])).toList();

  var ret = left
      .asMap()
      .entries
      .map((e) => (e.value * right.where((r) => r == e.value).length))
      .reduce((a, b) => a + b);

  print(ret);
}

void main() {
  String input = File('input.txt').readAsStringSync();
  List<String> lines = input.split('\n');
  part1(lines);
  part2(lines);
}
